src/filters/std/array.rs

```diff
@@ -101,6 +101,92 @@ impl Filter for SortNaturalFilter {
     }
 }
 
+#[derive(Debug, FilterParameters)]
+struct WhereArgs {
+    #[parameter(description = "The property being matched", arg_type = "str")]
+    property: Expression,
+    #[parameter(
+        description = "The value the property is matched with",
+        arg_type = "any"
+    )]
+    target_value: Option<Expression>,
+}
+
+#[derive(Clone, ParseFilter, FilterReflection)]
+#[filter(
+    name = "where",
+    description = "Filter the elements of an array to those with a certain property value. \
+                   By default the target is any truthy value.",
+    parameters(WhereArgs),
+    parsed(WhereFilter)
+)]
+pub struct Where;
+
+#[derive(Debug, FromFilterParameters, Display_filter)]
+#[name = "where"]
+struct WhereFilter {
+    #[parameters]
+    args: WhereArgs,
+}
+
+fn apply_where_filter<'a, T: Iterator<Item = &'a Object>>(
+    input: T,
+    property: &str,
+    target_value: Option<&Value>,
+) -> Vec<Value> {
+    match target_value {
+        None => input
+            .filter(|object| object.get(property).map_or(false, Value::is_truthy))
+            .map(|object| Value::Object(object.clone()))
+            .collect(),
+        Some(target_value) => input
+            .filter(|object| {
+                object
+                    .get(property)
+                    .map_or(false, |value| value == target_value)
+            })
+            .map(|object| Value::Object(object.clone()))
+            .collect(),
+    }
+}
+
+impl Filter for WhereFilter {
+    fn evaluate(&self, input: &Value, context: &Context) -> Result<Value> {
+        let args = self.args.evaluate(context)?;
+
+        match input {
+            Value::Array(array) => {
+                if array.is_empty() {
+                    // Retrun an empty array ony if we got an empty array
+                    return Ok(Value::array(vec![]));
+                }
+                let res = apply_where_filter(
+                    array.iter().filter_map(Value::as_object),
+                    &args.property,
+                    args.target_value,
+                );
+                if !res.is_empty() {
+                    Ok(Value::array(res))
+                } else {
+                    // We're required to return Nil if the original array wasn't
+                    // empty, but no element matched the filter. This is probably
+                    // an artifact of Ruby's select function.
+                    Ok(Value::nil())
+                }
+            }
+            // In the case of a single object, take it, but return an array.
+            Value::Object(object) => Ok(Value::array(apply_where_filter(
```

Comment:
The function is required to return `nil` in some cases (see `test_where_array_of_only_unindexable_values`), but you are correct that this isn't the proper condition. [The original Ruby implementation](https://github.com/Shopify/liquid/blob/98dfe198e168eef580345842ff49c4f2549e238b/lib/liquid/standardfilters.rb#L174) checks if the first element in the range is indexable, and return `nil` if that is not the case.
This has some weird implications:
* `where([{'a' => 1}, true])` throws, while  `where([true, {'a' => 1}])` returns `nil`.
* Since in Ruby ints are indexable (but only with ints), so  `where([1, {'a' => 1}], 'a')` throws. Interestingly enough,  `where([1, {'a' => 1}], 0)` returns [1].
Other filters (for example `sort` when given a property) check that all that all the elements are indexable, and if not it returns `nil`.

All this weirdness leaves us with two options:
* Try to be as compliant as we can, and check the first element for indexability, which will be defined in Ruby's terms.
* To make this function behave reasonably, and return `nil` only if we got an array that has a non-object element. A small side effect is that `where([nil], ...)` return `nil` while `where(nil, ...`) will fail.

I'm currently going for the second option unless you'll say otherwise.